### PR TITLE
Idle Connection Resiliency

### PR DESCRIPTION
**Pull request by v-zhangw**
Draft of Idle Connection Resiliency PR.
Additional documentation will be appended.

### CODE REVIEW

File: src/test/java/com/microsoft/sqlserver/jdbc/resiliency/ResultSetsWithResiliencyTest.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            insertData(s);
+        }
+    }
+
+    /*
+     * Execute statement with adaptive buffering on a broken connection.
+     */
+    @Test
+    public void testAdaptiveBuffering() throws SQLException {
+        verifyResulSetResponseBuffering("adaptive", true);
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/resiliency/ResultSetsWithResiliencyTest.java:46]: typo "ResulSet" - multiple occurrences

File: src/main/java/com/microsoft/sqlserver/jdbc/IdleConnectionResiliency.java [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+
+    void resetUnprocessedResponseCount() {
+        this.unprocessedResponseCount.set(0);
+    }
+
+    void parseInitialSessionStateData(byte[] data, byte[][] sessionStateInitial) throws SQLServerException {
+        int bytesRead = 0;
+
+        // Contains StateId, StateLen, StateValue
+        while (bytesRead < data.length) {
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/IdleConnectionResiliency.java:89]: null check data - this causes ArrayIndexOutofBoundsException exception against Azure db.

File: src/test/java/com/microsoft/sqlserver/jdbc/resiliency/BasicConnectionTest.java [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+            } catch (SQLException e) {
+                assertTrue(e.getMessage().contains("The connection is closed."));
+            }
+        }
+    }
+
+    @Test
+    public void testSetAttributes() throws SQLException {
+        try (Connection c = ResiliencyUtils.getConnection(connectionString)) {
+            ResiliencyUtils.toggleRandomProperties(c);
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/resiliency/BasicConnectionTest.java:59]: I think we need to separate these based on server support. e.g. SET REMOTE_PROC_TRANSACTIONS is not supported in Azure DB

Comment (lilgreenbird) [reply src/test/java/com/microsoft/sqlserver/jdbc/resiliency/BasicConnectionTest.java:61]: update - this is deprecated, we can remove this as per David

Comment (v-zhangw) [reply src/test/java/com/microsoft/sqlserver/jdbc/resiliency/BasicConnectionTest.java]: Removed.

File: src/test/java/com/microsoft/sqlserver/jdbc/resiliency/ResultSetsWithResiliencyTest.java (hunk 1 above)
Comment (v-zhangw) [reply src/test/java/com/microsoft/sqlserver/jdbc/resiliency/ResultSetsWithResiliencyTest.java]: Fixed 'resulset' typos.

**Pull request closed without merging**